### ISSUE TITLE
Cannot create new virtual disks in ESXi 6.7, 7 or 8 on a Datastore served over NFS from CephFS

### ISSUE THREAD

**Issue by karli-sjoberg** [nfs-ganesha/nfs-ganesha#1144]
After successfully adding a Datastore over NFSv 4.1 from RHCS 7 creating a new virtual machine in ESXi fails with error:

```
A general system error occurred: Error creating disk Device or resource busy
File system specific implementation of Ioctl[file] failed
File system specific implementation of Ioctl[file] failed
File system specific implementation of SetFileAttributes[file] failed
File system specific implementation of SetFileAttributes[file] failed
```

It has managed to create a file but looks it failed to set proper meta data:

```
drwxr-xr-x 2 root root 16G Jun  3 09:17 .
drwxrwxrwx 5 root root 37G Jun  3 09:31 ..
-rw------- 1 root root 16G Jan  1  1970 virtualmachine-flat.vmdk
```

This looks similar to what has been reported on your mailing list that files created are thick when they should be thin but I don't know whether that's the core issue or not:
[ESXI 6.7 client creating Thick Eager zeroed vmdk files using ceph fsal](https://lists.nfs-ganesha.org/archives/list/support@lists.nfs-ganesha.org/message/BQJWHS4ZUKO6KMY336RQLXS6M2A56EZU/)

Here is a full debug log of trying to create a new VM on this datastore:
[ganesha.tar.gz](https://github.com/user-attachments/files/15863797/ganesha.tar.gz)

Here are the share details:
```
# ceph nfs export ls nfs-ha --detailed                                     
[    
  {                       
    "access_type": "RW",
    "clients": [          
      {                   
        "access_type": "RW",
        "addresses": [
          "XXX.XXX.XXX.XXX",
          "XXX.XXX.XXX.XXX",
          "XXX.XXX.XXX.XXX",
          "XXX.XXX.XXX.XXX",
          "XXX.XXX.XXX.XXX",
          "XXX.XXX.XXX.XXX",
          "XXX.XXX.XXX.XXX",
          "XXX.XXX.XXX.XXX"
        ],               
        "squash": "no_root_squash"
      }    
    ],                 
    "cluster_id": "nfs-ha",
    "export_id": 1,
    "fsal": {                   
      "cmount_path": "/",
      "fs_name": "test",
      "name": "CEPH",
      "user_id": "nfs.nfs-ha.test"
    },
    "path": "/",    
    "protocols": [        
      4                      
    ],           
    "pseudo": "/test",
    "security_label": false,
    "squash": "no_root_squash",
    "transports": [   
      "TCP"
    ]
  }
]

# ceph nfs export info nfs-ha /test                                        
{
  "access_type": "RW",
  "clients": [
    {
      "access_type": "RW",
      "addresses": [
          "XXX.XXX.XXX.XXX",
          "XXX.XXX.XXX.XXX",
          "XXX.XXX.XXX.XXX",
          "XXX.XXX.XXX.XXX",
          "XXX.XXX.XXX.XXX",
          "XXX.XXX.XXX.XXX",
          "XXX.XXX.XXX.XXX",
          "XXX.XXX.XXX.XXX"
      ],
      "squash": "no_root_squash"
    }
  ],
  "cluster_id": "nfs-ha",
  "export_id": 1,
  "fsal": {
    "cmount_path": "/",
    "fs_name": "test",
    "name": "CEPH",
    "user_id": "nfs.nfs-ha.test"
  },
  "path": "/",
  "protocols": [
    4
  ],
  "pseudo": "/test",
  "security_label": false,
  "squash": "no_root_squash",
  "transports": [
    "TCP"
  ]
}
```

Here is the service specification for NFS:
```
service_type: nfs
service_id: nfs-ha
service_name: nfs.nfs-ha
placement:
  count: 1
  hosts:
  - ceph-gw005
  - ceph-gw006
spec:
  port: 12049
```

And here's the specification for the HAProxy/Keepalived:
```
service_type: ingress
service_id: nfs.nfs-ha
service_name: ingress.nfs.nfs-ha
placement:
  count: 1
  hosts:
  - ceph-gw005
  - ceph-gw006
spec:
  backend_service: nfs.nfs-ha
  frontend_port: 2049
  monitor_port: 9049
  virtual_ip: XXX.XXX.XXX.XXX
```

Ceph is version 18 and Ganesha is version 5.6.
  
A test was set up to first copy a file with random data of 10 MB in size from ESXi-host "hv12" onto the datastore and then tried  to create a new virtual machine.

Before performing the test I started `tcpdump` on the ESXi-host like so:
`# tcpdump-uw -s0 -i vmk0 -W 10 -C 512 -w hv12.pcap -Z root port 2049 and host XXX.XXX.XXX.XXX`

I started capture on the HAProxy/Keepalived like:
`# tcpdump -s0 -W 10 -C 512 -w ceph-gw005.pcap -Z root`

Like this on the NFS server:
`# tcpdump -s0 -W 10 -C 512 -w ceph-gw006.pcap -Z root`

I then performed the test and stopped the captures. Copying the file succeeded but creating a new VM failed. Here are the packet captures:
[ceph-gw005.pcap0.tar.gz](https://github.com/user-attachments/files/15864157/ceph-gw005.pcap0.tar.gz)
[ceph-gw006.pcap0.tar.gz](https://github.com/user-attachments/files/15864161/ceph-gw006.pcap0.tar.gz)
[hv12.pcap0.tar.gz](https://github.com/user-attachments/files/15864163/hv12.pcap0.tar.gz)

Thanks in advance!

**Comment by dang** [comment]
Thanks for the detailed bug report!  I took a look at the ganesha log and at the client pcap.  The log doesn't include any WRITE ops at all, so something must be wrong with the collection of the log.  In the pcap, there is a write of a file named "test.bin", that succeeds, including the follow-up SETATTR ops.  There isn't any write that I can find of the vmdk file, so I can't determine what went wrong there.   (VMWare is a very weird implementation of NFS, or maybe just it's use of NFS, so pcaps from VMWare clients are extremely chatty and hard to follow).

Can you double check that the logs/pcaps were from the same time period as the attempted creation of the vmdk?  Also, I'd like to get Frank's opinion on this, but he's out of the office until Wed.

**Comment by karli-sjoberg** [comment]
Yes it seems the Ganesha logs weren't fully there, apologies. I redid the test procedure while debug was on and some of the log files ended up way bigger so I'll have to link to them from my Nextcloud (yes the cert is self signed, sue me :wink: )

[ceph-gw005-20240618-1.pcap0.tar.gz](https://cloud.inparadise.se:8443/f/219704)
[ceph-gw006-20240618-1.pcap0.tar.gz](https://cloud.inparadise.se:8443/f/219709)
[hv12-20240618-1.pcap0.tar.gz](https://cloud.inparadise.se:8443/f/219705)
[ganesha-20240618-1.tar.gz](https://cloud.inparadise.se:8443/f/219707)

The name of the VM I tried to create was called "virtual-machine" to make searching a little easier.

Thanks in advance!

**Comment by dang** [comment]
Your nextcloud is asking me for a login.   Is there a no-login link I could get?  Or you could email me a temporary login directly, I guess.

**Comment by karli-sjoberg** [comment]
My bad, wrong links:

[ceph-gw005-20240618-1.pcap0.tar.gz](https://cloud.inparadise.se:8443/s/PwYxpjZysn9Z4KH)
[ceph-gw006-20240618-1.pcap0.tar.gz](https://cloud.inparadise.se:8443/s/pEpPDPFjeAEq3tq)
[ganesha-20240618-1.tar.gz](https://cloud.inparadise.se:8443/s/ci9zxmgsxDxmsp4)
[hv12-20240618-1.pcap0.tar.gz](https://cloud.inparadise.se:8443/s/Q2gyM7tGkFq2tyG)

The previous ones were internal links which is why it was prompting for login.

**Comment by karli-sjoberg** [comment]
Have you been able to download the archives now?

**Comment by dang** [comment]
I got them, yes.

**Comment by dang** [comment]
Sorry, I have upcoming vacation, and have things I need to get done for work.  Maybe Frank can look in the interim, or I'll get to this when I'm back.

**Comment by ffilz** [comment]
I tried to fetch ganesha-20240618-1.tar.gz and got a file named: ci9zxmgsxDxmsp4 (which is right in the link). Renaming it to ganesha-20240618-1.tar.gz didn't work. How do I actually get this?

What do you mean by thick vs. thin files? FSAL_CEPH doesn't fully support sparse files (the SEEK operation and READ_PLUS are not supported though it does seem to support ALLOCATE and DEALLOCATE. And creating a file with a size set SHOULD create a sparse file.

**Comment by karli-sjoberg** [comment]
@ffilz You have to click and follow the link to able to download the archive, I tested it just now to be sure.

Thanks in advance!
Karli Sjöberg

**Comment by lknite** [comment]
I'm running into the exact same errors.  I can mount the nfs share via linux and create files, I can upload files to the nfs share mounted as a datastore, but when I try to create a vm I get the same errors:
```
A general system error occurred: Error creating disk Device or resource busy
File system specific implementation of Ioctl[file] failed
File system specific implementation of Ioctl[file] failed
File system specific implementation of SetFileAttributes[file] failed
File system specific implementation of SetFileAttributes[file] failed
```

ceph deployed into kubernetes via rook-ceph helm charts v1.15.0
vcenter 8u3
ceph version 18.2.4 (e7ad5345525c7aa95470c26863873b581076945d) reef (stable)
ganesha version 5.5

**Comment by lknite** [comment]
does this mean there is no one using ceph/nfs with the vcenter/vmware setup? ... surprising

**Comment by lknite** [comment]
@dang back from vacation, might i tempt you to look at this one again?

I'm thinking it might be time to switch to proxmox ...

**Comment by karli-sjoberg** [comment]
@dang @ffilz

Have you had time to analyze the info I sent you about this issue yet?

**Comment by PredatarIan** [comment]
FWIW i've had some slightly better success by using nfs-ganesha directly on the host rather then the ceph orch container version:
- apt install libcephfs2 nfs-ganesha nfs-ganesha-ceph
- cp /etc/ganesha/ceph.conf /etc/ganesha/ganesha.conf
- modify ganesha.conf for your settings
- systemctl start nfs-ganesha

**Comment by patrakov** [comment]
Maybe because this is a different version?

My observation: nfs-ganesha 4.0.8 works, 5.9 fails to create new VM disks, succeeds to copy them from other storage, fails to start the imported VM (VMware says "insufficient resources" with no information at all on what exactly is insufficient).

**Comment by lknite** [comment]
@patrakov could you provide a work around?  Were you perhaps simply able to specify a different image version via the rook helm chart somewhere to cause an older version of nfs-ganesha to be used?  Could you provide the helm configuration (related to nfs) you use so that I could use it as a work around until this issue is resolved?

**Comment by patrakov** [comment]
I deployed a non-containerized NFS-Ganesha 4.0.8 by building it from source, on a host which has nothing else.

**Comment by karli-sjoberg** [comment]
@dang @ffilz

I'd appreciate some kind of feedback here! If anything just say no and close the issue 🤷

**Comment by RobertLukan** [comment]
I am facing the same problem. I am on 19.2 that was just released. I can copy files from another volume to CEPH NFS volume. I have created the cluster by using cephadm orch, with ingress feature enabled. I have tried thin and thick provisioned. I noticed that problem is when vmdk is getting created and not other files such as vmx. I would like to stay with nfs v4 if possible. Any ideas ?

**Comment by kasim457** [comment]
> I deployed a non-containerized NFS-Ganesha 4.0.8 by building it from source, on a host which has nothing else.

Could you pls share steps to build  non-containerized NFS-Ganesh 4.0.8

**Comment by karli-sjoberg** [comment]
@dang @ffilz

I'm wondering if you could please take a look at the debug-information I sent you nearly half a year ago and comment if this is even something you're going to do so I can stop pestering you for updates!

Best Regards

**Comment by ffilz** [comment]
Could you try on Ganesha V6.3? I see a bunch of SETATTR failing with SHARE_DENIED in the log. I'm not quite sure what's going wrong but it would help to make sure it's not something already fixed by some of the file handling changes that have gone in since V5.6.

**Comment by ffilz** [comment]
Also, the truth is while our team has grown, it is still a pretty small team and supporting users outside support contracts is a lower priority for us. We make a best effort, but we aren't super human.

And anything that's a real bug or has potential is work keeping open in case others hit the same issue, or we have cycles to address it.

What would also help is if you are able to do some debug on your side and find a small re-create that we can run.

p.s. I see that the client is doing OPEN ACCESS_BOTH, DENY_ALL. Linux clients don't use DENY modes so it's possible some logic it broken there that we don't have an easy test for (though I thought pynfs does some testing of these scenarios).

**Comment by patrakov** [comment]
I have retested with NFS-Ganesha 6.3. Nope, VMware still cannot use disks. I will try to bisect which commit broke it.

**Comment by patrakov** [comment]
Bisected, although using a slightly different definition of "good," a revision is good if a pre-created VM can be started from the NFS41 datastore.

Bisection result for a VM backed by a directory on FSAL_CEPH:

```
7c2374e78c876645afc38befcfbd7f7756b51488 is the first bad commit
commit 7c2374e78c876645afc38befcfbd7f7756b51488
Author: Frank S. Filz <ffilzlnx@mindspring.com>
Date:   Mon Feb 7 11:57:34 2022 -0800

    FSAL_CEPH: Use new fsal_fd management
    
    Change-Id: Ifdf11a1bad429229b808038ab016682f06c351b7
    Signed-off-by: Frank S. Filz <ffilzlnx@mindspring.com>

 src/FSAL/FSAL_CEPH/handle.c   | 1013 +++++++++++++++++++----------------------
 src/FSAL/FSAL_CEPH/internal.c |    5 +
 src/FSAL/FSAL_CEPH/internal.h |    6 +-
 3 files changed, 477 insertions(+), 547 deletions(-)
```

For a VM backed by a local directory and FSAL_VFS, everything works with 6.3. The VM can be started and a new disk can be added.

**Comment by kasim457** [comment]
@patrakov   I have created a new docker image with nfs-ganesha-6.3 rpm packages and I have not yet tested it.  Do you still problem in creation of new VM's with nfs-ganesh 6.3 packages?

**Comment by ffilz** [comment]
I'm really not sure how to proceed here. We need a re-create that doesn't depend on vmware in order to debug what exactly is going wrong.

**Comment by patrakov** [comment]
Hello Frank,

I have already cloned PyNFS, but that's it. I know where to create a test file, and I have looked at simple tests that already exist there. However, I am a slow coder and I am swamped with other work, so, even if I could, I wouldn't have time to write a test case.

What I _can_ do is provide a pcap of a one good and one bad session. I also have a few commits that happened before the fsal_fd management rework that, according to their descriptions, might be relevant. I will try reverting them or moving the offending commit as far back in the history as it might be possible and bisecting that range.

**Comment by kasim457** [comment]
@patrakov  @ffilz   Any further updates from your team on NFS access issue ?

**Comment by hcbwiz** [comment]
Is it related to fallocate?

**ceph: only allow punch hole mode in fallocate**
https://lore.kernel.org/lkml/87sh1evyii.fsf@suse.com/T/

https://github.com/ceph/ceph/pull/59725

**Comment by lknite** [comment]
Just a note here, I abandoned vmware in preference for proxmox, because of this issue.

**Comment by bolduant** [comment]
Hi, any updates so far?

**Comment by ffilz** [comment]
No, we need someone who can debug the situation with ESXi or find a recreate that does not depend on that.

**Comment by bolduant** [comment]
Maybe I can do the debugging,because I'm in the exact same situation. What exactly do you need?

**Comment by ffilz** [comment]
Well, the ideal would be "here's the line of code causing the problem and here's the patch"...

More helpfully, what would really help is enough information to try and make a re-create so it can be debugged with debug logging and using gdb to look at memory.

But it may not be quite that bad, so at least a short tcpdump trace that shows the sequence of ops that lead to the error. Knowing what op resulted in the error, and what ops might have been involved in setting up for that op might make the issue pop out in clarity to then find the offending code.

**Comment by bolduant** [comment]
Thank you for the info. I will do my best.
Here is a tcpdump from the vmware side:

[nfs.pcap.zip](https://github.com/user-attachments/files/19436621/nfs.pcap.zip)

Hope that helps.


it happens first time at packet nr. 483-484

**Comment by bolduant** [comment]
hi, any updates ont this?
Do you need anything else I should provide?

**Comment by ffilz** [comment]
OK, I see the problem, in ceph_fsal_setattr2, we check for share conflict when we are changing file size AND we have a state.

We should check if we DON'T have a state...

If we have a state, we SHOULD check that the open mode is compatible with setting size.

If you are up for trying to fix this, that would help, otherwise I will try and get to it, but right now I have a bunch of things on my plate.

It's only a problem because you have a client that actually uses deny modes...

**Comment by bolduant** [comment]
Thank you. I would try to fix if you tell me how I can do this. Is there
any patch?

Frank Filz ***@***.***> schrieb am Di., 8. Apr. 2025, 20:57:

**Comment by ffilz** [comment]
I don't have a patch yet... Need time to write code, or have someone else take a stab at it...

**Comment by bolduant** [comment]
I maybe found the place where to change the code, in handle.c in folder FSAL_CEPH on line 2792 there starts the check for "changing file size AND having a state", I don't really know if it should be replaced or if you should just add additional code for the "check if we DON'T have a state"-thing, but I think this snipped will do it:

```
        /* Test if size is being set, make sure file is regular and if so,
	 * require a read/write file descriptor.
	 */
	if (state == NULL &&
	    FSAL_TEST_MASK(attrib_set->valid_mask, ATTR_SIZE)) {
		if (obj_hdl->type != REGULAR_FILE) {
			LogFullDebug(COMPONENT_FSAL,
				     "Setting size on non-regular file");
			return fsalstat(ERR_FSAL_INVAL, EINVAL);
		}

		/* Now check the new share and establish if OK. */
		status = check_share_conflict_and_update_locked(
			obj_hdl, &myself->share, FSAL_O_CLOSED, FSAL_O_RDWR,
			false);

		if (FSAL_IS_ERROR(status)) {
			LogFullDebug(
				COMPONENT_FSAL,
				"check_share_conflict_and_update_locked status=%s",
				fsal_err_txt(status));
			goto out;
		}

		need_share = true;
	}
```

maybe that helps

**Comment by rrmichel** [comment]
@ffilz 
> It's only a problem because you have a client that actually uses deny modes...

I'm not sure...but where is the "deny" defined? I'm at a loss right now.

**Comment by ffilz** [comment]
@bolduant Yea, that's a good start.

We should also somewhere, if a stateid is provided for SETATTR, make sure that the stateid allows the proposed changes, for example, you need to be open for write to set the size.

**Comment by ffilz** [comment]
> [@ffilz](https://github.com/ffilz)
> 
> > It's only a problem because you have a client that actually uses deny modes...
> 
> I'm not sure...but where is the "deny" defined? I'm at a loss right now.

NFSv4 OPEN specifies an access mode (read, write, read/write) and a deny mode (deny none, deny read, deny write, deny read/write aka deny both). The deny mode matches Windows open semantics (well, dating back to DOS 4). NFSv3 locking, NLM, actually provides an NLM_SHARE operation that attempts the same thing though I/O is not tied to an NLM_SHARE so it can't actually be enforced (but at least a DOS/Windows NFS client can't open a file in a conflicting mode to another DOS/Windows NFS client).

**Comment by sajibreadd** [comment]
Can anyone explain the state here? NFS v4.1 is stateful no? Is vmware doing something fishy such that we can get state is NULL. 
Sorry I am just a beginner in this NFS Ganesha product. My company assigned me to somehow getting it fixed. So if someone can share any information, which might help me understand the code. That would be great.

**Comment by bolduant** [comment]
> Can anyone explain the state here? NFS v4.1 is stateful no? Is vmware doing something fishy such that we can get state is NULL. Sorry I am just a beginner in this NFS Ganesha product. My company assigned me to somehow getting it fixed. So if someone can share any information, which might help me understand the code. That would be great.

Hi Sajib,
I'm, unfortunately, also not into nfs ganesha but as far as I understood the code, the errors from VMware and the messages in this bug, VMware trys to set attributes, even if they don' have a state. As you mentioned, the state is NULL; As you might have seen in the message from @ffilz "ceph_fsal_setattr2, we check for share conflict when we are changing file size AND we have a state.".

In my post I tried to find the position for this check and changed/added it to a version where a state can be NULL, but I'm not a programmer and I do not know how to build nfs ganesha with the changes I added.

**Comment by sajibreadd** [comment]
@bolduant safe approach would be just finding the code flow for state==NULL before the devil commit(`7c2374e78c876645afc38befcfbd7f7756b51488`) happenned. And replicate that code flow for this new version. I will try that and post it here.

**Comment by patrakov** [comment]
@sajibreadd Please do not ignore the fact that the VFS FSAL works. Maybe we should, in addition to examining the bisection result, find analogies between the two FSALs.

**Comment by bolduant** [comment]
@sajibreadd thank you for all your effort.
If I can do any I will, but I mostly do not understand any of what you wrote ^^

**Comment by sajibreadd** [comment]
@ffilz I did some investigation on this issue. Here is the approach I am following to debug this as I don't have much understading regarding the `share` mechanism. I took two commit branch `7c2374e78c876645afc38befcfbd7f7756b51488` (vmware doesn't work) and it's parent commit `9a8136292df7cd359cf5782da062c02ad5b7d1e1` (vmware works for it).

My goal was to understanding the code workflow.
For commit `7c2374e78c876645afc38befcfbd7f7756b51488` in ceph_fsal_setattr2 function
```
if (state != NULL &&
	    FSAL_TEST_MASK(attrib_set->valid_mask, ATTR_SIZE)) {
		if (obj_hdl->type != REGULAR_FILE) {
			LogFullDebug(COMPONENT_FSAL,
				     "Setting size on non-regular file");
			return fsalstat(ERR_FSAL_INVAL, EINVAL);
		}
		struct fsal_fd *state_fd = (struct fsal_fd *) (state + 1);
		LogInfo(COMPONENT_FSAL,
		     "state_fd->openflags = %d openflags = %d",
		     state_fd->openflags, FSAL_O_RDWR); // this log added by me

		/* Now check the new share and establish if OK. */
		status = check_share_conflict_and_update_locked(obj_hdl,
								&myself->share,
								FSAL_O_CLOSED,
								FSAL_O_RDWR,
								false);
		LogInfo(COMPONENT_FSAL,
		     "ceph_fsal_setattr2: access_read=%u, access_write=%u, deny_read=%u, deny_write=%u, deny_write_v4=%u, open_correct(state_fd->openflags, FSAL_O_RDWR)=%d",
		     myself->share.share_access_read,
		     myself->share.share_access_write,
		     myself->share.share_deny_read,
		     myself->share.share_deny_write,
		     myself->share.share_deny_write_mand,
			open_correct(state_fd->openflags, FSAL_O_RDWR)); // this log added by me

		if (FSAL_IS_ERROR(status)) {
			LogFullDebug(COMPONENT_FSAL,
				     "check_share_conflict_and_update_locked status=%s",
				     fsal_err_txt(status));
			goto out;
		}

		need_share = true;
	}
```

I found this log, 
```
11/07/2025 11:36:01 : epoch 6870daa0 : sajibreadd : ganesha.nfsd-1280158[svc_4] ceph_fsal_setattr2 :FSAL :INFO :ceph_fsal_setattr2: access_read=1, access_write=1, deny_read=1, deny_write=1, deny_write_v4=1, open_correct(state_fd->openflags, FSAL_O_RDWR)=1
```
I am logging open_correct function's output (why am I doing it I will explain in the next phase)
please note that `deny_write=1 and deny_read=1`

For commit `9a8136292df7cd359cf5782da062c02ad5b7d1e1`
Same piece of code but was done in a different way,
```
if (FSAL_TEST_MASK(attrib_set->valid_mask, ATTR_SIZE)) {
		if (obj_hdl->type != REGULAR_FILE) {
			LogFullDebug(COMPONENT_FSAL,
				     "Setting size on non-regular file");
			return fsalstat(ERR_FSAL_INVAL, EINVAL);
		}

		/* We don't actually need an open fd, we are just doing the
		 * share reservation checking, thus the NULL parameters.
		 */
		status = fsal_find_fd(NULL, obj_hdl, NULL, &myself->share,
				      bypass, state, FSAL_O_WRITE, NULL, NULL,
				      &has_lock, &closefd, false,
				      &reusing_open_state_fd);
		struct fsal_fd *state_fd = (struct fsal_fd *) (state + 1);
		LogInfo(COMPONENT_FSAL,
		     "ceph_fsal_setattr2: access_read=%u, access_write=%u, deny_read=%u, deny_write=%u, deny_write_v4=%u, open_correct(state_fd->openflags, FSAL_O_RDWR)=%d",
		     myself->share.share_access_read,
		     myself->share.share_access_write,
		     myself->share.share_deny_read,
		     myself->share.share_deny_write,
		     myself->share.share_deny_write_mand,
			open_correct(state_fd->openflags, FSAL_O_RDWR));

		if (FSAL_IS_ERROR(status)) {
			LogFullDebug(COMPONENT_FSAL,
				     "fsal_find_fd status=%s",
				     fsal_err_txt(status));
			goto out;
		}
	}
```

I found this log,
```
11/07/2025 11:47:02 : epoch 6870dd54 : sajibreadd : ganesha.nfsd-1292682[svc_9] ceph_fsal_setattr2 :FSAL :INFO :ceph_fsal_setattr2: access_read=1, access_write=1, deny_read=1, deny_write=1, deny_write_v4=1, open_correct(state_fd->openflags, FSAL_O_RDWR)=1
```

So in both case `deny_read=1, deny_write=1`, so far we are good.

Now let's talk about the part where 2 versions are working differentlly
In latest version it called `check_share_conflict_and_update_locked` directly but parent version it did the same thing with `fsal_find_fd` function. But there is a catch.

Inside `fsal_find_fd` function of the parent version has a piece of code
```
fsal_status_t fsal_find_fd(struct fsal_fd **out_fd,
			   struct fsal_obj_handle *obj_hdl,
			   struct fsal_fd *obj_fd,
			   struct fsal_share *share,
			   bool bypass,
			   struct state_t *state,
			   fsal_openflags_t openflags,
			   fsal_open_func open_func,
			   fsal_close_func close_func,
			   bool *has_lock,
			   bool *closefd,
			   bool open_for_locks,
			   bool *reusing_open_state_fd)
{
	fsal_status_t status = {ERR_FSAL_NO_ERROR, 0};
	struct fsal_fd *state_fd;
	struct fsal_fd *related_fd;
	struct state_t *openstate;

	if (state == NULL)
		goto global;

	/* Check if we can use the fd in the state */
	state_fd = (struct fsal_fd *) (state + 1);

	LogFullDebug(COMPONENT_FSAL,
		     "state_fd->openflags = %d openflags = %d%s",
		     state_fd->openflags, openflags,
		     open_for_locks ? " Open For Locks" : "");

	if (open_correct(state_fd->openflags, openflags)) {
		/* It was valid, return it.
		 * Since we found a valid fd in the state, no need to
		 * check deny modes.
		 */
		LogFullDebug(COMPONENT_FSAL, "Use state_fd %p", state_fd);
		if (out_fd)
			*out_fd = state_fd;
		*has_lock = false;
		return status;
	}
```
Now please attention on the open_correct function which is providing true, it just returns from there, so it doesn't check the `check_share_conflict` thing which comes later. Also from the comment you can see `no need to check deny modes`

But in the latest version, we are calling `check_share_conflict_and_update_locked` function unconditionally, even if `open_correct` function is returning true.

In my opinion we should skip the `check_share_conflict_and_update_locked` if `open_correct` is true.

Let me know your thoughts. Also how can we contribute to the code, Can we just create pull request?

**Comment by sajibreadd** [comment]
Hi @ffilz 
I know you have some DevPolicy for patches. But I am creating this pull request as a discussion base. It would be great if you can share some feedback.
I would really appreciate some "WHY" question from you while reviewing.

**Comment by ffilz** [comment]
Please submit via gerrithub per src/CONTRIBUTING_HOWTO.txt

**Comment by sajibreadd** [comment]
@ffilz Just did it https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/1218357

**Comment by patrakov** [comment]
I confirm that the change of `state != NULL` to `state == NULL` on top of NFS-Ganesha 6.5 allows NFS 4.1 datastore to function correctly in VMware ESXi 7.0.3 and 8.0.3.

(sorry, I cannot comment on gerrithub, the registration never finishes, probably because I only requested reviewer access and I am not in the EU)

**Comment by lknite** [comment]
definitely recommend switching to proxmox and using the built-in ceph support

**Comment by patrakov** [comment]
Why is this still open? The commit was merged long ago.

**Comment by ffilz** [comment]
Case of not going back and closing resolved issues after patches have been merged... :-)

**Comment by ffilz** [comment]
Also, originators are welcome to close issues they consider resolved.
